### PR TITLE
Performance improvement in cudf::strings::find/rfind for long strings

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Improves performance for longer strings with `cudf::strings::find()` and `cudf::strings::rfind()` APIs.

The current implementation works well with small-ish strings and so this new implementation splits into a long-ish string algorithm when the average number of bytes per string is 64 bytes or greater. The new implementation searches for the target string by applying a warp per string kernel. Additionally, the special logic needed for matching an empty target (easily checked in host code) is factored out into its own transform functor.
For longer strings, the performance improvement is about 2-6x.

Reference https://github.com/rapidsai/cudf/issues/13048

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
Some benchmark results from `benchmarks/STRINGS_BENCH --benchmark_filter=StringFindScalar/find/`

| rows | width | baseline | warp | x |
|:---:|:---:| ---:| ---:| ---:|
| 4096  | 32     | 0.037  |  0.037 |  1.00  |
| 4096  | 64     | 0.045  |  0.045 |  1.00  |
| 4096  | 128    | 0.062  |  0.062 |  1.00  |
| 4096  | 256    | 0.102  |  0.040 |  2.55  |
| 4096  | 512    | 0.169  |  0.048 |  3.52  |
| 4096  | 1024   | 0.292  |  0.061 |  4.79  |
| 4096  | 2048   | 0.463  |  0.084 |  5.51  |
| 4096  | 4096   | 0.802  |  0.141 |  5.69  |
| 4096  | 8192   |  1.40  |  0.238 |  5.88  |
| 32768  | 32    | 0.040  |  0.040 |  1.00  |
| 32768  | 64    | 0.048  |  0.047 |  1.02  |
| 32768  | 128   | 0.065  |  0.065 |  1.00  |
| 32768  | 256   | 0.106  |  0.074 |  1.43  |
| 32768  | 512   | 0.197  |  0.108 |  1.82  |
| 32768  | 1024  | 0.326  |  0.153 |  2.13  |
| 32768  | 2048  | 0.524  |  0.240 |  2.18  |
| 32768  | 4096  | 0.890  |  0.415 |  2.14  |
| 32768  | 8192  |  1.48  |  0.766 |  1.93  |
| 262144  | 32   | 0.061  |  0.061 |  1.00  |
| 262144  | 64   | 0.091  |  0.091 |  1.00  |
| 262144  | 128  | 0.177  |  0.177 |  1.00  |
| 262144  | 256  | 0.788  |  0.339 |  2.32  |
| 262144  | 512  |  1.40  |  0.558 |  2.51  |
| 262144  | 1024 |  2.08  |  0.882 |  2.36  |
| 262144  | 2048 |  5.75  |   1.45 |  3.97  |
| 262144  | 4096 |  15.3  |   2.69 |  5.69  |
| 2097152  | 32  | 0.215  |  0.214 |  1.00  |
| 2097152  | 64  | 0.403  |  0.402 |  1.00  |
| 2097152  | 128 | 0.941  |  0.941 |  1.00  |
| 2097152  | 256 |  6.40  |   2.45 |  2.61  |
| 2097152  | 512 |  9.94  |   4.16 |  2.39  |
| 16777216  | 32 |  1.44  |   1.43 |  1.01  |
| 16777216  | 64 |  2.87  |   2.87 |  1.00  |

File: cpp/src/strings/search/find.cu [hunk 1]
```diff
@@ -41,103 +41,207 @@ namespace cudf {
...
- * @brief Utility to return integer column indicating the position of
- * target string within each string in a strings column.
+ * @brief Threshold to decide on using string or warp parallel functions.
  *
- * Null string entries return corresponding null output column entries.
+ * If the average byte length of a string in a column exceeds this value then
+ * a warp-parallel function is used.
  *
- * @tparam FindFunction Returns integer character position value given a string and target.
+ * Note that this value is shared by find, rfind, and contains functions
```
Comment (ttnghia) [inline cpp/src/strings/search/find.cu:51]:
```suggestion
 * Note that this value is shared by find, rfind, and contains functions.
```

File: cpp/src/strings/search/find.cu [hunk 2]
```diff
@@ -41,103 +41,207 @@ namespace cudf {
...
+ * @brief Find function handles a string per thread
+ */
+template <bool forward = true>
+struct finder_fn {
+  column_device_view const d_strings;
+  string_view const d_target;
+  size_type const start;
+  size_type const stop;
+
+  __device__ size_type operator()(size_type idx)
```
Comment (ttnghia) [inline cpp/src/strings/search/find.cu:65]:
```suggestion
  __device__ size_type operator()(size_type idx) const
```

File: cpp/src/strings/search/find.cu [hunk 3]
```diff
@@ -41,103 +41,207 @@ namespace cudf {
...
  */
-template <typename FindFunction>
-std::unique_ptr<column> find_fn(strings_column_view const& strings,
+template <bool forward = true>
+struct empty_target_fn {
+  column_device_view const d_strings;
+  size_type const start;
+  size_type const stop;
+
+  __device__ size_type operator()(size_type idx)
```
Comment (ttnghia) [inline cpp/src/strings/search/find.cu:93]:
```suggestion
  __device__ size_type operator()(size_type idx) const
```

**Review by ttnghia — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
Sorry for the delay. It took me ages to understand the new kernel.

**Pull request merged**